hw9: advanced oop/hw9.py

```diff
@@ -0,0 +1,212 @@
+# this homework explores the functionality of object-oriented programming in Python #
+#####################################################################################
+
+# task 1: connected classes Chat, Message and User
+
+import datetime
+
+
+class Chat:
+    def __init__(self):
+        self.chat_history = []
+
+    def show_last_message(self):
+        # the assumption is made that we live in the one timeline
+        # therefore, the last message is one sent as the latest
+        last_message = self.chat_history[0]
+        return last_message.show()
+
+    def get_history_from_time_period(self, start, stop):
+        chat_sorted = Chat()  # type(self) ?
+
+        for message in self.chat_history[::-1]:
+            if start <= message.datetime <= stop:
+                chat_sorted.recieve(message)
+
+        return chat_sorted
+
+    def show_chat(self):
+        for message in self.chat_history:
+            message.show()
+            print()
+
+    def recieve(self, message):
+        self.chat_history = [message] + self.chat_history
+
+
+class Message:
+    def __init__(self, text, date_time, user, ability='is a god/ess'):
+        self.text = text
+        self.datetime = datetime.datetime.strptime(date_time, "%d/%m/%y %H:%M")
+        self.user = User(ability=ability, name=user)
+
+    def show(self):
+        print(self.datetime, self.user.name, self.text)
+
+    def send(self, chat):
+        chat.recieve(self)
+
+
+class User:
+    def __init__(self, name, ability):
+        self.ability = ability
+        self.name = name
+
+    def greeting(self):
+        print(f'Hi, {self.name}! U r gorgeous')
+
+    def description(self):
+        print(f'{self.name} {self.ability}')
+
+
+# task 2: new syntax for Python
+
+# to avoid annoying nested list let's flatten them
+def flatten_list(nested_list):
+    flattened_list = []
+    for element in nested_list:
+        if isinstance(element, list):
+            flattened_list += flatten_list(element)
+        else:
+            flattened_list.append(element)
+    return flattened_list
+
+
+class Args:
+    def __init__(self, *args, **kwargs):
+        self.args = list(args)
+        self.args = flatten_list(self.args)
+        self.kwargs = list(kwargs.values())
+
+        self.all_argies = self.args + self.kwargs
+
+    def __rlshift__(self, other):
+        return other(*self.all_argies)
+
+
+# task 3: a new way to pass arguments to functions
+
+class StrangeFloat(float):
+    def __getattr__(self, attr):
+        if '__' + attr.split('_')[0][:3] + '__' in dir(float):
+            fnc = getattr(float, '__' + attr.split('_')[0][:3] + '__')
+            return fnc(self, float(attr.split('_')[1]))
+        else:
+            dict_attr = {'subtract': self - float(attr.split('_')[1]),
+                         'multiply': self * float(attr.split('_')[1]),
+                         'divide': self / float(attr.split('_')[1])}
+            return StrangeFloat(dict_attr[attr.split('_')[0]])
```

Comment:
Очень трудно читается, `attr.split('_')` повторяется аж **7 раз!** Давай упростим

```suggestion
    def __getattr__(self, attr):
        command, value = attr.split("_")
        value = float(value)
        dict_attr = {'add': self.__add__,
                          'subtract': self.__sub__,
                          'multiply': self.__mul__,
                          'divide': self.__truediv__}
        result = dict_attr[command](value)
        return StrangeFloat(result)
```